ERCS/erc-7546.md

```diff
@@ -0,0 +1,178 @@
+---
+eip: 7546
+title: Upgradeable Clone for Scalable Contracts
+description: An upgradeable, cloneable, horizontally extensible proxy pattern.
+author: Shogo Ochiai (@shogochiai) <shogo.ochiai@pm.me>, Kai Hiroi (@KaiHiroi) <kai.hiroi@pm.me>
+discussions-to: https://ethereum-magicians.org/t/eip-7546-upgradeable-clone/16256
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-10-25
+requires: 165, 1967, 7201
+---
+
+## Abstract
+It has been a significant challenge for developers attempting to create cloneable and upgradeable contracts on the Ethereum Virtual Machine (EVM). While UUPS and the Diamond Standard offer partial solutions, a comprehensive answer has remained elusive. Our proposal addresses this gap through the introduction of two main features.
+
+### Function-Level Upgradeability
+In alignment with [ERC-2535](./eip-2535.md), this functionality permits the selective redirection of implementation contracts for individual function calls. This granular control over upgrades allows for modifications on a per-function basis. Moreover, segmenting implementation contracts by function helps mitigate the limitations posed by the contract size cap (24.576kB as of EVM version Shanghai or earlier).
+
+### Factory/Clone-Friendly & Simultaneous Upgradeability
+Drawing on the Beacon model from [ERC-1967](./eip-1967.md), our method aims to streamline the process of cloning and updating Proxy contracts simultaneously. This approach is designed to maintain consistent functionality across different instances, each with its own state. Typically, proxies are limited to basic upgradeability features or follow the [ERC-1167](./eip-1167.md) standard. However, our solution combines both functionalities into a compact proxy, developed using the Huff language.
```

Comment:
> developed using the Huff language

Is Huff necessary to implement your proposal, or is this just background information? Remember that you're writing a standard that (hopefully) many others will want to implement, and not just a single implementation. Try to keep things general and avoid implementation details in this document (except for the Reference Implementation section; implementation specifics are fine there.)

Reply:
Understood. I'll remove the implementation details from this section.